python/src/iceberg/table/schema.py

```diff
@@ -0,0 +1,286 @@
+        raise NotImplementedError()
+
+
+class SchemaVisitor(Generic[T], ABC):
+    def before_field(self, field: NestedField) -> None:
+        pass
+
+    def after_field(self, field: NestedField) -> None:
+        pass
+
+    def before_list_element(self, element: NestedField) -> None:
+        self.before_field(element)
+
+    def after_list_element(self, element: NestedField) -> None:
+        self.after_field(element)
+
+    def before_map_key(self, key: NestedField) -> None:
+        self.before_field(key)
+
+    def after_map_key(self, key: NestedField) -> None:
+        self.after_field(key)
+
+    def before_map_value(self, value: NestedField) -> None:
+        self.before_field(value)
+
+    def after_map_value(self, value: NestedField) -> None:
+        self.after_field(value)
+
+    @abstractmethod
+    def schema(self, schema: Schema, struct_result: T) -> T:
+        ...
+
+    @abstractmethod
+    def struct(self, struct: StructType, field_results: List[T]) -> T:
+        ...
+
+    @abstractmethod
+    def field(self, field: NestedField, field_result: T) -> T:
+        ...
+
+    @abstractmethod
+    def list(self, list_type: ListType, element_result: T) -> T:
+        ...
+
+    @abstractmethod
+    def map(self, map_type: MapType, key_result: T, value_result: T) -> T:
+        ...
+
+    @abstractmethod
+    def primitive(self, primitive: PrimitiveType) -> T:
+        ...
+
+
+def visit(obj, visitor: SchemaVisitor[T]) -> T:
+    if isinstance(obj, Schema):
+        return visitor.schema(obj, visit(obj.as_struct(), visitor))
+
+    elif isinstance(obj, StructType):
+        results = []
+        for field in obj.fields:
+            visitor.before_field(field)
+            try:
+                result = visit(field.type, visitor)
+            finally:
+                visitor.after_field(field)
+
+            results.append(visitor.field(field, result))
+
+        return visitor.struct(obj, results)
+
+    elif isinstance(obj, ListType):
+        visitor.before_list_element(obj.element)
+        try:
+            result = visit(obj.element.type, visitor)
+        finally:
+            visitor.after_list_element(obj.element)
+
+        return visitor.list(obj, result)
+
+    elif isinstance(obj, MapType):
+        visitor.before_map_key(obj.key)
+        try:
+            key_result = visit(obj.key.type, visitor)
+        finally:
+            visitor.after_map_key(obj.key)
+
+        visitor.before_map_value(obj.value)
+        try:
+            value_result = visit(obj.value.type, visitor)
+        finally:
+            visitor.after_list_element(obj.value)
+
+        return visitor.map(obj, key_result, value_result)
+
+    elif isinstance(obj, PrimitiveType):
+        return visitor.primitive(obj)
+
+    else:
+        raise NotImplementedError("Cannot visit non-type: %s" % obj)
+
+
+def index_by_id(schema_or_type) -> Dict[int, NestedField]:
+    class IndexById(SchemaVisitor[Dict[int, NestedField]]):
```

Comment:
Does nesting this class redefine it each time? We could always move it out. I don't see a big problem with exposing it under `schema`.

Reply:
Moved this out to the module level and added a `_` prefix to mark it as internal to the module.